### ISSUE TITLE
2.x -- group by, custom map variant

### ISSUE THREAD

**Issue by sfitts** [ReactiveX/RxJava#5763]
We are using release 2.1.7.

As is noted in the "What's Different" page, the variant of `groupBy` that accepts a custom map factory is not currently part of the 2.x API.  We were relying on this to implement "eviction" of groups that haven't been seen for a "long time" (where "long" is user specified).  Our system maintains Flowable instances for its entire lifetime (which is hopefully weeks/months), so it is important to be able to reclaim resources as we go.

Anyway -- my questions are -- are the plans to bring this back?  Is there a suggested alternative for implementing the disposal of groups after a period of inactivity?  Any guidance would be appreciated.

**Comment by akarnokd** [comment]
It was a community contribution which was not ported to 2.x by the original author. You can timeout groups and switch to an empty flow:

```java
source.groupBy(v -> key(v))
.map(g -> g.timeout(5, TimeUnit.MINUTES, Observable.empty())
.flatMap(g -> /* the usual group processing */)
```

**Comment by sfitts** [comment]
Thanks, that should work nicely.

**Comment by davidmoten** [comment]
I was author of that contribution to 1.x #3931. Using `timeout` was discussed but it had problems. In particular you can lose emissions. If @akarnokd is in favour I can investigate adapting the `groupBy` operator to use a Map factory again.

**Comment by akarnokd** [comment]
Yes, PR welcome.

**Comment by sfitts** [comment]
@davidmoten thanks for the info re: timeout dropping emissions.  We have essentially the same use case that you describe in #3931, so something that avoids that issue for Rx2 would be great.

**Comment by davidmoten** [comment]
Just looking at the PR now. @akarnokd do you have any advice for optimizing the RxJava 2.x compile experience in Eclipse (latest)? I just waited 5 min 40s for the project to compile after modifying some code and this happens frequently. 

PS Just realized I'm not using the Eclipse Gradle plugin for the project in Eclipse but rather running `./gradlew eclipse` and importing as an Eclipse project. What do you do @akarnokd?

**Comment by akarnokd** [comment]
None. The bug has been reported and hopefully we get it backported to 4.7.

**Comment by davidmoten** [comment]
Ok. You still using Eclipse for dev on RxJava?

**Comment by akarnokd** [comment]
Yes. I just do something else for 10 minutes while Eclipse is compiling. After that, as long as you don't touch the main classes, it's okay.

**Comment by davidmoten** [comment]
Eclipse slow compilation bug:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=521438

**Comment by vanniktech** [comment]
I'm using IntelliJ and it works without any problems. Also Android studio picks RxJava up correctly.

**Comment by davidmoten** [comment]
Thanks @vanniktech, I might switch to that for RxJava contributions.

**Comment by akarnokd** [comment]
Closing via #5860.

**Comment by davidmoten** [comment]
Woohoo! Slow compilation of RxJava 2.x is fixed in Eclipse 4.7.3.
